### PR TITLE
Review Changes in Development Branch

### PR DESCRIPTION
**Pull request by smk508**
Key Changes:

- Simplified naming scheme and syntax
- Broke down Sources into Pipes and Junctions. Pipes have a single input, which makes recursive method calling an unambiguous task. Junctions can have multiple inputs (called components), and you can write arbitrary logic to aggregate these components.
- Added Model class for representing mathematical models. These models function like Pipes and can be put into a data pipeline. They also function like Junctions in that their parameters are components, and these Parameters can be trained using PyTorch.

### CODE REVIEW

File: Fireworks/database.py [hunk 1]
```diff
@@ -3,19 +3,19 @@
...
     """
-    Represents an SQLalchemy Table while having the functionality of a Source.
+    Represents an SQLalchemy Table while having the functionality of a Pipe.
     """
-    def __init__(self, table, engine, columns = None, inputs = None, **kwargs):
+    def __init__(self, table, engine, columns = None, input = None, **kwargs):
 
-        super().__init__(inputs=inputs, **kwargs)
+        super().__init__(input=input, **kwargs)
         Session = sessionmaker(bind=engine)
```
Comment (haq204) [inline Fireworks/database.py]: I don't think it's a good idea to open a sqlalchemy session inside the class instance. Typically you want session management outside and independent of data operations. Also I do don't see a method to close a session, which leaves the risk of having open threads. Consider making session an input to TableSource because the TableSource shouldn't be managing the session or if you want to keep engine an input have the engine execute the query directly like

```
connection = engine.connect()
result = connection.execute("select username from users")
for row in result:
    print("username:", row['username'])
connection.close()
```

File: Fireworks/database.py [hunk 2]
```diff
@@ -186,21 +186,21 @@ def __repr__(self):
 
     return SimpleTable
```
Comment (haq204) [inline Fireworks/database.py]:
```
class SimpleTable(Base):
        __table__ = Table(name, Base.metadata, *columns)
```
This can cause some unexpected bugs or behavior. You define Base above with declarative_base. which this SimpleTable will inherit. The problem is that it will be mapped to THAT specific base class. If you call it again it will give you warning that a similar table has already been mapped. Consider making Base an input and let the user control which base class to use.

File: Fireworks/database.py [hunk 3]
```diff
@@ -186,21 +186,21 @@ def __repr__(self):
...
         Args:
             table (sqlalchemy.ext.declarative.api.DeclarativeMeta): Table to perform query on. You can alternatively provide the name of the
                 table as a string, and the schema will be extracted from the engine.
             engine (sqlalchemy.engine.base.Engine): Engine correspondign to the database to read from.
-            query: Can optionally provide an SQLalchemy Query object. If unspecified, the DBSource will perform a SELECT * query.
+            query: Can optionally provide an SQLalchemy Query object. If unspecified, the DBPipe will perform a SELECT * query.
         """
+        Pipe.__init__(self)
         self.engine = engine
         Session = sessionmaker(bind=engine)
```
Comment (haq204) [inline Fireworks/database.py]: Same here DBPipe shouldn't be managing the session let that be an input when initiating or query using the engine directly.

File: Fireworks/database.py (hunk 1 above)
Comment (haq204) [reply Fireworks/database.py]: from the sqlalchemy docs 
```
tl;dr;

As a general rule, keep the lifecycle of the session separate and external from functions and objects that access and/or manipulate database data. This will greatly help with achieving a predictable and consistent transactional scope.

Make sure you have a clear notion of where transactions begin and end, and keep transactions short, meaning, they end at the series of a sequence of operations, instead of being held open indefinitely.
```

File: Fireworks/model.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        return self.forward(message)
+
+    def __next_hook(self, message):
+
+        return self.forward(message)
+
+    def __call__(self, message, *args, **kwargs):
+
+        try: # This will trigger a recursive call if possible.
+            message = self.input(message, *args, **kwargs)
```
Comment (haq204) [inline Fireworks/model.py]: I think from a readability standpoint define this with a different variable like `recursive_message` or something. You should also probably define this at the beginning with `None` and maybe do a type check to make it more clear

**Comment by haq204** [comment]
This is a really monster PR. Can you cherrypick and break break this up into smaller PRs? Also there seems to be a lot of unnecessary files in VCS, further making it harder to review. Thanks

File: Fireworks/database.py (hunk 2 above)
Comment (smk508) [reply Fireworks/database.py]: Base is now an input that the user can provide, and if not provided, the function creates a new declarative_base instance. The SQLalchemy warnings are gone now.

File: Fireworks/extensions/database.py [hunk 5]
```diff
@@ -44,14 +42,17 @@ def rollback(self):
         """
         self.session.rollback()
 
+    def reset_session(self, session=None):
```
Comment (haq204) [inline Fireworks/extensions/database.py]: I would rename this to `create_session`

File: Fireworks/extensions/database.py [hunk 6]
```diff
@@ -2,22 +2,21 @@
...
-from Fireworks.source import Source, PassThroughSource
+from Fireworks import Message, cat, Pipe
 import numpy as np
 import pandas as pd
 
-Base = declarative_base()
+#TODO: Develop a way to infer schemas from Messages and Pipes to make save/load easier.
 
-class TableSource(PassThroughSource):
+class TablePipe(Pipe):
```
Comment (haq204) [inline Fireworks/extensions/database.py]: Also add a method to close a session

File: Fireworks/extensions/database.py [hunk 7]
```diff
@@ -2,22 +2,21 @@
...
     """
-    def __init__(self, table, engine, columns = None, inputs = None, **kwargs):
+    def __init__(self, table, engine, columns = None, input = None, **kwargs):
 
-        super().__init__(inputs=inputs, **kwargs)
-        Session = sessionmaker(bind=engine)
-        self.session = Session()
+        super().__init__(input=input, **kwargs)
+        self.sessionmaker = sessionmaker(bind=engine)
+        self.reset_session()
```
Comment (haq204) [inline Fireworks/extensions/database.py]: don't create a session at instance. Make it so that the use has to explicitly define when a session begins

File: Fireworks/extensions/database.py [hunk 8]
```diff
@@ -186,22 +191,22 @@ def __repr__(self):
 
     return SimpleTable
 
-class DBSource(Source):
+class DBPipe(Pipe):
```
Comment (haq204) [inline Fireworks/extensions/database.py]: If DBPipe is supposed to represent the query result shouldn't the actual query be performed in TablePipe? DBPipe should be agnostic of context and should take in just the query result. Because you're creating two sessions to do a query when you don't have to. It's unclear which session is actually performing the query.

File: Fireworks/database.py (hunk 2 above)
Comment (haq204) [reply Fireworks/database.py]: this looks fine

File: Fireworks/extensions/database.py (hunk 8 above)
Comment (smk508) [reply Fireworks/extensions/database.py]: As far as I can tell, sessions and queries are tightly connected in sqlalchemy (https://docs.sqlalchemy.org/en/latest/orm/query.html). What's happening here, is you can make any SQLalchemy query on a TablePipe using the query method, and the result of that query will be used to make a DBPipe. The query object has a session, and that session also has its own associated query method. In theory, I could desession the DBPipe by manually setting query.session = None, and then requiring the user to provide a session, however, I think the approach here works fine. I can change it slightly to enforce that the session associated with a query is the same as self.session to remove ambiquity.
Both DBPipe and TablePipe have a session attribute, and manipulation of that session is separate from the methods that manipulate data. Hence, you can still swap out sessions easily and make your own decisions about how to manage them without having to mess with those methods.

File: Fireworks/extensions/database.py (hunk 5 above)
Comment (haq204) [reply Fireworks/extensions/database.py]: fair enough

File: Fireworks/extensions/database.py (hunk 8 above)
Comment (haq204) [reply Fireworks/extensions/database.py]: You only need to provide a session to the query object if you're instantiating yet directly (`Query(entities, session)`). You can also do `session.query(*args)`. which will make a query object with that session and is probably the preferred way.  It seems like the best way to handle this is instead of DBPipe taking in an engine, it takes in instead a session (that will have an engine associated anyway). That way when you query in TablePipe it passes in the TablePipe session into DBPipe. So instead of 
```DBPipe(self.table, self.engine, query, columns_and_types = columns_and_types)```
it's 
```DBPipe(self.table, self.session, query, columns_and_types = columns_and_types)```

This makes it more clear and better lines up the way sqlalchemy works at the ORM level

File: Fireworks/core/pipe.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
+from abc import ABC, abstractmethod
```
Comment (haq204) [inline Fireworks/core/pipe.py]: I don't see most of these imports used. Why are they still here?

File: Fireworks/core/pipe_test.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
+import os
```
Comment (haq204) [inline Fireworks/core/pipe_test.py]: some of these imports aren't used so not sure why they are still here.

File: Fireworks/core/pipe.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+
+    def __iter__(self, *args, **kwargs):
+        return self.input.__iter__(*args, **kwargs)
+
+    def __getattr__(self, *args, **kwargs):
+        """
+        Pass through all methods of the input Pipe while adding labels. This does not intercept special methods (__x__ methods)
+        """
+
+        return self.recursive_call(*args, **kwargs) #self.input.__getattribute__(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (smk508) [reply Fireworks/core/pipe.py]: This is a typo. I'll fix that.

File: Fireworks/extensions/database.py (hunk 8 above)
Comment (smk508) [reply Fireworks/extensions/database.py]: I think I'm going to keep self.engine for now, because I realized that DBPipe attempts to infer column names from a table using table reflection if those are not provided by the user. This function requires an engine. 
https://docs.sqlalchemy.org/en/latest/core/reflection.html 
This is a convenience feature which lets you provide the name of a table as a string and a query, and the object will reflect the table out from the database and resolve the query. We can think about how to simplify this module and make it more streamlined and generalized, but I don't think it's a high priority right now because it works.

**Comment by smk508** [comment]
> It seems like you've basically define what args are so why not make them actual variables? also, you're pulling out specific things in kwargs so why not also just make them inputs?
> 
> ```
> class Message:
> ...
> def __init__(self, *args, metadata = None, length = None, **kwargs):
>         """
>         Initializes a message
>         If *args is empty --> empty message
>         If *args is two args --> construct a message assuming arg0 is tensors and arg1 is dataframe
>         Alternatively, tensors and dataframes can be specified with kwargs (tensors  = ... and df = ...)
>         If *args has __getitem__ and keys() attributes --> construct a message from elements of dict, separating tensors from pd series
>         """
> ```
> ```
> if 'tensors' in kwargs:
>             tensors = kwargs['tensors']
>         if 'df' in kwargs:
>             df = kwargs['df']
> ```

The initializer is set up that way so that it behaves differently depending on the quantity and type of arguments.
For example, if you do 
`m = Message(A)
`if will treat A as a dictionary and separate out the tensor and non-tensor components into a DataFrame and TensorMessage. On the other hand, if you do
`m = Message(A, B)
`it will treat A as a dictionary that should be mapped to self.TensorMessage and B as a dictionary that should be mapped to self.df. It will attempt to convert elements of A and B in order to enforce this.

Alternatively, you can provide these arguments via keyword. The initializer is set up to behave differently in each of these situations by parsing the size or args and kwargs.

**Comment by smk508** [comment]
> What's the purpose of being able to input a Message to create a Message? Also how do you handle column name ambiguity in a Message between the TensorMessage part and pandas DataFrame part?
> 
> ```
> if len(args) == 1:
>             # Identify tensors and separate them out
>             # The argument may be an existing Message/TensorMessage
>             # TODO: Implement ability to create a message from a list of messages/dicts
>             if type(args[0]) is Message:
>                 tensors = args[0].tensor_message
>                 df = args[0].df
> ```

That is so that the initializer is idempotent. You can call 
`m = Message(Message(x))
`without raising errors. This is useful because if you have a function that expects arguments to be Messages, it can enforce this by attempting to convert its argument into a Message. If it's already a Message, then this check will cause the rest of the initializer to be skipped and return the argument, thus eliminating some computations. On the other hand, if a user provides a dataframe or something, it will be converted on the fly, and this extends the flexibility of the function.

You can't have overlapping keys in a Message. If you try to do that, the values will overwrite, and the key will point to whatever value was last specified. You should not do this.

**Comment by haq204** [comment]
> You can't have overlapping keys in a Message. If you try to do that, the values will overwrite, and the key will point to whatever value was last specified. You should not do this.

Can we add a check for that then?

File: Fireworks/extensions/database.py (hunk 8 above)
Comment (haq204) [reply Fireworks/extensions/database.py]: maybe then make sessionmaker an additional input, because you don't need two sessionmakers (if it's used in conjunction with TablePipe). It also gives the user more control what kind of sessions they want to use (ex. Thread local sessions).

**Comment by haq204** [comment]
> Alternatively, you can provide these arguments via keyword. The initializer is set up to behave differently in each of these situations by parsing the size or args and kwargs.

Does the documentation say what those keywords are?

File: Fireworks/core/message.py [hunk 12]
```diff
@@ -349,9 +314,47 @@ def __contains__(self, item):
...
+            return self.df.__getattribute__(name)
+        elif name in self.tensor_message.keys():
+            return self.tensor_dict.__getattr__(name)
+        else:
+            raise AttributeError("This message has no attribute or column named {0}.".format(name))
+
     def __repr__(self):
         return "Message with \n Tensors: \n {0} \n Metadata: \n {1}".format(self.tensor_message, self.df)
 
+    def to(self, method, *args, **kwargs):
```
Comment (haq204) [inline Fireworks/core/message.py]: can you make it more explicit what are args and kwargs? As the saying goes, friends don't let other friends use kwargs :)
